packages/libs/error-stack/src/fmt/hook.rs

```diff
@@ -0,0 +1,594 @@
+    _marker: PhantomData<T>,
+}
+
+impl<T> HookContext<'_, T> {
+    /// If [`Debug`] requests, this text (can include line breaks) will be appended to the main
+    /// message.
+    /// This is especially useful for dense information like backtraces,
+    /// which one might want to omit from the tree rendering or omit from the normal debug.  
+    pub fn text(&mut self, value: &str) {
+        self.text_lines(value.lines());
+    }
+
+    /// Same as [`Self::text`], but only accepts lines (the internal representation used during
+    /// rendering)
+    pub fn text_lines(&mut self, lines: core::str::Lines) {
+        self.parent
+            .text
+            .push(lines.map(ToOwned::to_owned).collect());
+    }
+}
+
+impl<'a, T> HookContext<'a, T> {
+    /// Cast the `HookContext` to a new type.
+    ///
+    /// This binds the parent context to a different value,
+    /// which allows to access the data stored for the type this is casted to.
+    #[must_use]
+    pub fn cast<U>(self) -> HookContext<'a, U> {
+        HookContext {
+            parent: self.parent,
+            _marker: PhantomData::default(),
+        }
+    }
+
+    #[cfg(feature = "hooks")]
+    fn into_impl(self) -> &'a mut HookContextImpl {
+        self.parent
+    }
+}
+
+impl<T: 'static> HookContext<'_, T> {
+    /// Returns a reference to the value corresponding to the currently bound type `T`.
+    ///
+    /// This will downcast the stored value to `U`, if not possible, it will return [`None`]
+    #[must_use]
+    pub fn get<U: 'static>(&self) -> Option<&U> {
+        let inner = self.parent.inner.get(&TypeId::of::<T>())?;
+
+        inner.downcast_ref()
+    }
+
+    /// Returns a mutable reference to the value corresponding to the currently bound type `T`.
+    ///
+    /// This will downcast the stored value to `U`, if not possible, it will return [`None`]
+    pub fn get_mut<U: 'static>(&mut self) -> Option<&mut U> {
+        let inner = self.parent.inner.get_mut(&TypeId::of::<T>())?;
+
+        inner.downcast_mut()
+    }
+
+    /// Insert a value into the context of the currently bound type `T`.
+    ///
+    /// Returns the previously stored value, downcast to `U`, if no previous value was stored, or
+    /// the downcast was not possible will return [`None`]
+    pub fn insert<U: 'static>(&mut self, value: U) -> Option<Box<U>> {
+        let inner = self
+            .parent
+            .inner
+            .insert(TypeId::of::<T>(), Box::new(value))?;
+
+        inner.downcast().ok()
+    }
+
+    /// One of the most common interactions with [`HookContext`] is a counter to reference previous
+    /// frames or the content emitted during [`text()`].
+    ///
+    /// This is a utility method, which uses the other primitive methods provided to automatically
+    /// increment a counter, if the counter wasn't initialized this method will return `0`.
+    ///
+    /// [`text()`]: Self::text
+    pub fn incr(&mut self) -> isize {
```

Comment:
Can we have an example here on what `incr` and `decr` is doing? This isn't intuitive from only reading the documentation.